kernel/src/transaction/mod.rs

```diff
@@ -508,17 +509,23 @@ impl Transaction {
     // Generate the logical-to-physical transform expression which must be evaluated on every data
     // chunk before writing. At the moment, this is a transaction-wide expression.
     fn generate_logical_to_physical(&self) -> Expression {
-        // for now, we just pass through all the columns except partition columns.
-        // note this is _incorrect_ if table config deems we need partition columns.
         let partition_columns = self
             .read_snapshot
             .table_configuration()
             .metadata()
             .partition_columns();
         let schema = self.read_snapshot.schema();
+
+        // Check if materializePartitionColumns feature is enabled
+        let materialize_partition_columns = self
+            .read_snapshot
+            .table_configuration()
+            .protocol()
+            .has_table_feature(&TableFeature::MaterializePartitionColumns);
+
         let fields = schema
             .fields()
-            .filter(|f| !partition_columns.contains(f.name()))
+            .filter(|f| materialize_partition_columns || !partition_columns.contains(f.name()))
```

Comment:
mega nit: i spent a second thinking about this - maybe a quick comment on what's going on